### PR TITLE
Add a signature validation budget during path construction

### PR DESCRIPTION
**Pull request by woodruffw**
This extends our existing NC budget check to include a budget for signature validations. If a path construction exceeds the budget by performing more than the allowed number of signature validation steps, the entire construction fails.

For now, our budget is 128 signature validations. This is consistent with (higher than) Go and rustls-webpki, which both set a limit of 100. Like Go, we attempt to make the "best" use of our signature budget by ordering by likelihood, using AKI/SKI match as the strongest signal of fitness.

### CODE REVIEW

File: src/rust/cryptography-x509-verification/src/lib.rs [hunk 1]
```diff
@@ -528,4 +595,102 @@ mod tests {
...
+        fn clone_public_key(_key: &Self::Key) -> Self::Key {}
+
+        fn clone_extra(_extra: &Self::CertificateExtra) -> Self::CertificateExtra {}
+    }
+
+    #[test]
+    fn test_clone() {
+        assert_eq!(NullOps::clone_public_key(&()), ());
+        assert_eq!(NullOps::clone_extra(&()), ());
+    }
```
Comment (woodruffw) [inline src/rust/cryptography-x509-verification/src/lib.rs:631]: We need `NullOps` for the coverage below, which means we need to "test" the empty trait impl bodies we add.

(We already do this for another test-only impl, `PulicKeyErrorOps`.)

File: src/rust/cryptography-x509-verification/src/lib.rs [hunk 2]
```diff
@@ -528,4 +595,102 @@ mod tests {
...
+        let mut budget = Budget {
+            name_constraint_checks: usize::MAX,
+            signature_checks: usize::MAX,
+        };
+
+        let name_chain = NameChain::new::<NullOps>(None, &ca_exts, false)
+            .ok()
+            .unwrap();
+        let err = builder
+            .build_chain_inner(&working, u8::MAX, &ca_exts, name_chain, &mut budget)
```
Comment (woodruffw) [inline src/rust/cryptography-x509-verification/src/lib.rs:688]: This single line is the coverage-bearing one.

**Review by alex — APPROVED** [review]
